ui/app/helpers/utils/i18n-helper.js

```diff
@@ -39,13 +40,41 @@ export const getMessage = (localeCode, localeMessages, key, substitutions) => {
   }
   const entry = localeMessages[key]
   let phrase = entry.message
+
+  const requiresWrappers = phrase.match(/\{\{.+?\}\}/)
+
+  const hasSubstitutions = Boolean(substitutions && substitutions.length)
+  const hasReactSubstitutions = hasSubstitutions && substitutions.some((element) => typeof element === 'function' || typeof element === 'object')
+
   // perform substitutions
-  if (substitutions && substitutions.length) {
+  if (hasReactSubstitutions || requiresWrappers) {
+    const parts = phrase.split(/(\$\d|\{\{.+?\}\})/g)
+
+    const wrappedAndSubStitutedParts = parts.map((part) => {
+      if (part.match(/\{\{.+?\}\}/)) {
+
+        const wrapper = reactWrappers.shift()
+        return React.cloneElement(
+          wrapper,
+          wrapper.props,
+          [ part.match(/[^{}]+/) ]
+        )
+      } else if (part.match(/\$\d/)) {
+        return substitutions.shift()
+      } else {
+        return part
+      }
+    })
+
+    phrase = React.createElement('span', null, ...wrappedAndSubStitutedParts)
+  } else if (hasSubstitutions) {
```

Comment:
It looks like this being in the `else if` block would preclude mixing React substitutions and string substitutions. is that necessary? I don't see why we would disallow mixing those.

Reply:
Oh, I see that you're handling substitutions in the above block as well. I'm guessing that was because `phrase` becomes a React component in that block so it wasn't straightforward to continue modifying the string message.

I think this could be simplified by doing normal substitutions first, then scanning the resulting string for React substitutions. That way the logic for doing the normal substitutions only has to be in one place

Reply:
Simplification as suggested done in 247b12834